Issue 46 units selection

juzen2003 opened the pull request:
- Fixes #46 and #886 
- Merge pull request https://github.com/SETI/pds-opus/pull/891 as the base, and update from there.
- Run JSHint on: search.js, hash.js, and widgets.js
- Test on latest Chrome, FF, Opera, and Safari.
- Done Items:
    - Update unit- to opus.extras in the following cases: 
        - Open a widget and select a different value in qtype:
            - properly update opus.extras, unit should stay in unit- not qtype-
            - properly update opus.selections/selections, unit should stay in extras with unit- not selections
        - Properly add unit- in URL when open a widget
        - Properly remove unit- in URL when close a widget
        - Update unit- in URL when an input is added/remove 
        - Update unit- in URL when refreshing the page with multiple input in a widget 
        - Update unit- in URL when calling API (getHashStrFromSelections) 
        - Update unit- when page reloads (getHashStrFromSelections, initFromHash) 
    - Update hints when unit is changed.
    - No search will be performed if unit is changed when all input sets are empty, only hints will be updated.
    - No search will be performed if qtype is changed when corresponding input set is empty.

rfrenchseti committed:
Snapshot of Django units support

rfrenchseti committed:
Clean up search routines. More unit work. Code coverage test updates. Update windows requirements.

rfrenchseti committed:
Add units to normalizeurl and update tests

rfrenchseti committed:
More normalizeurl tests for units

rfrenchseti committed:
Fix bug in unit conversion

rfrenchseti committed:
Update units

rfrenchseti committed:
Update units

rfrenchseti committed:
Change range/endpoints to take units parameter.

rfrenchseti committed:
Merge branch 'django_units' of https://github.com/rfrenchseti/pds-opus into django_units

rfrenchseti committed:
Remove trailing zeros from floating point metadata output values. Fixes #886. Adjust floating point format for scale of unit.

rfrenchseti committed:
Fix fields.csv and unit tests

rfrenchseti committed:
normalizeurl enforces same unit for all clauses of a slug

rfrenchseti committed:
Merge branch 'fix_normalizeurl' into django_units

juzen2003 committed:
Merge remote-tracking branch 'rob_repo/django_units' into 46_units_selection

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 46_units_selection

juzen2003 committed:
Fix the unit conversion factor for km/pixel in opus.support.py. Fix tests
in test_ui_api.py. Put recyclebin in SLUGS_NOT_IN_DB.
- In opus_support.py (line 476), change from "1e3" to "1e-3".
- In settings.py (line 304), put "recyclebin" to SLUGS_NOT_IN_DB.
- In test_ui_api.py (line 2375, 2381, 2387), fix tests.

juzen2003 committed:
Add select input of units next to the widget header if the widget has units
available.
- In views.py (line 321,334), pass valid_units to widget.html.
- In widget.html (line 14-22), add select input of units next to widget
  header if units are available in the widget.

juzen2003 committed:
Update unit-slug to opus.extras in the following functions: initFromHash,
getSelectionsExtrasFromHash.
- In hash.js (line 528), update unit-slug to opus.extras in initFromHash.
- In hash.js (line 379), update unit-slug to opus.extras in getSelectionsExtrasFromHash.
- In hash.js (line 623-734), properly update unit-slug to opus.extras in
  alignDataInSelectionsAndExtras.
- In hash.js (line 737), create syncUpExtrasWithEmptySelections to sync
  up extras with selections when slug in selections is empty but qtype-slug
  or unit-slug in extras exists, we will also make sure data in selections
  and extras are aligned.
- In opus.js, add debugging console.log.

juzen2003 committed:
Update unit-slug in URL when an input is added/remove, refreshing the page
with multiple inputs in a widget, and calling API.
- In hash.js (line 107-111,136-140,145-183), modify getHashStrFromSelections
  to properly update hash string with unit-slug in opus.extras.
- In search.js (line 370,379), widgets.js (line 939,944,957,1113,1179),
  add .widget-main to qtype input selector (to be more specific) so that
  it won't select the unit input.
- In widgets.js (line 334-363), properly remove unit-slug in opus.extras
  when an input is removed.
- In widgets.js (line 261-277), properly add unit-slug in opus.extras when
  an input is added.

juzen2003 committed:
Update unit-slug when a widget is open/closed, update result count when
a unit input is changed.
- In hash.js (line 533-541), modify extrasWithoutUnusedQtypes to include
  unit-slug for extras comparison.
- In search.js (line 363-396), update select change event handler. Update
  unit-slug to opus.extras when unit input is changed.
- In widgets.js (line 941-954), modify getWidget (done callback function)
  to update unit-slug to opus.extras when a widget is open.
- In widgets.js (line 616), remove unit-slug in opus.extras when a widget
  is closed.

juzen2003 committed:
Update hints when a unit is changed in a widget.
- In search.js (line 897-906), modify getRangeEndpoints to add "units" to
  range endpoints api url if unit is available.

juzen2003 committed:
Update select change event handler to make sure no search is performed
when (1) qtype is changed but the corresponding input is empty and (2)
unit is changed but the corresponding input sets are all empty (only update
hints in this case).
- In search.js (line 363-440), update select change event handler to have
  the behavior mentioned above.
- In search.js (line 444-458), create isSlugSelectionsEmpty to check if
  slugSelections is an array of null to determine if all input sets are
  empty.

juzen2003 committed:
Clean up unwanted console.log and commented out codes.

juzen2003 committed:
Clean up commented out codes in widgets.js.

rfrenchseti commented on log_analyzer/opus/slug.py:
```diff
@@ -70,7 +70,7 @@ class ToInfoMap:
     # Slugs that should be ignored when see them as either a column name or as a search term.
     SLUGS_NOT_IN_DB = {'browse', 'cart_browse', 'cart_startobs', 'col_chooser', 'colls_browse', 'cols', 'detail',
                        'gallery_data_viewer', 'limit', 'loc_type', 'order', 'page',
-                       'range', 'recyclebin', 'reqno', 'request', 'startobs', 'types', 'url_cols', 'view', 'widgets', 'widgets2',
```
There appears to be a merge error with this. The recyclebin and url_cols slugs need to stay in the list.

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -11,7 +11,14 @@
-                {{ label }} {{ units }}
+                {{ label }}&nbsp;
+                {% if valid_units %}
+                    <select class="unit-{{ slug }}" name="unit-{{ slug }}" tabindex="0">
```
Shouldn't these names of "op-" in front of them? And do we really need identical classes and IDs?

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -126,38 +131,59 @@ var o_hash = {
 
-                    if (qtypeValue.length > 1) {
+                    if (slugInExtrasVal.length > 1) {
                         let numberOfQtypeInputs = encodedExtraValues.length;
```
Shouldn't this be just `numberOfInputs` or something since this can either be qtype or unit slug?

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -126,38 +131,59 @@ var o_hash = {
+                            if (slugInExtrasVal[trailingCounter-1] !== null) {
                                 hash.push(newKey + "=" + encodedExtraValues[trailingCounter-1]);
                             }
+                            if (unitSlug in opus.extras) {
```
This is kind of tricky. I finally figured out that "unitSlug" is empty if the main slug is already a unit slug, which is counter-intuitive. I think this needs some kind of comment here and line 177, or perhaps just above where unitSlug is first created.

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -126,38 +131,59 @@ var o_hash = {
+                            let newKey = `${slugInExtras}_${trailingCounterString}`;
 
-                            if (qtypeValue[trailingCounter-1] !== null) {
+                            if (slugInExtrasVal[trailingCounter-1] !== null) {
```
Why is the "main" slug (whether qtype or unit) inserted directly using hash.push, but the associated unit slug, if any, is inserted by calling `updateHashFromExtras`? Shouldn't they be consistent?

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -354,7 +383,7 @@ var o_hash = {
         }
 
-        hash = (hash.search('&') > -1 ? hash.split('&') : [hash]);
+        hash = (hash.search("&") > -1 ? hash.split("&") : [hash]);
```
Comment before this function should say "qtype or unit"

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -419,7 +448,7 @@ var o_hash = {
     },
 
-    SyncUpWithOPUSSelectionsForWidgetsWithNoQtype: function(selections, extras, widgetSlug) {
+    syncUpWithOPUSSelectionsForWidgetsWithNoQtype: function(selections, extras, widgetSlug) {
```
I probably understood this function once, but I don't anymore. Can you elaborate in the comments why this function here and what it does? Under what circumstances is there a string/range widget but no qtype?

rfrenchseti replied:
Also, below we check for qtype in extras. Do we also need to check for unit-? Is the name of the function correct?

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -522,7 +554,7 @@ var o_hash = {
 
             if (value) {
-                if (slug.match(/qtype-.*/)) {
+                if (slug.match(/qtype-.*/) || slug.match(/unit-.*/)) {
```
Better with startsWith?

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -360,38 +360,99 @@ var o_search = {
+                let counterStr = o_utils.getSlugOrDataTrailingCounterStr($(this).attr("name"));
+                let idx = counterStr ? counterStr - 1 : 0;
+                switch ($(this).attr("class")) {  // form type
+                    case "RANGE":
```
I'm not sure why this wasn't done before, but the code after each "case" should be indented one level.

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -360,38 +360,99 @@ var o_search = {
+                let units = [];
+                let slugNoNum = $(this).attr("name").match(/unit-(.*)$/)[1];
+                let numberOfInputSets = $(`#widget__${slugNoNum} .op-search-inputs-set`).length;
+                while(units.length < numberOfInputSets) {
```
Space after while

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -591,6 +601,7 @@ var o_widgets = {
 
         delete opus.extras[`qtype-${slugNoNum}`];
         delete opus.extras[`z-${slugNoNum}`];
+        delete opus.extras[`unit-${slugNoNum}`];
```
Let's get rid of that line that says "z-". That's left over from some ancient feature we no longer support. Perhaps it was the old version of units? Before my time.

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -909,15 +920,29 @@ var o_widgets = {
-            let qtype = "qtype-" + slug;
-            let qtypeInputs = $(`#widget__${slug} select[name="${qtype}"]`);
+            let qtype = `qtype-${slug}`;
+            let qtypeInputs = $(`#widget__${slug} .widget-main select[name="${qtype}"]`);
```
Should .widget-main be renamed to .op-widget-main?

rfrenchseti reviewed (changes requested):
- Nicely done. I only found one real issue and a handful of small code comments.
- Preprogrammed ranges have units associated with them. Right now they are ignored. But when the user selects a range, the UI needs to check if the ranges' unit is equal to the widget's currently selected unit. If not, and there are other input lines with values in them, there needs to be some kind of error dialog. If there are no other lines with values in them, then the widget's unit needs to be set appropriately.
- Unrelated to the current pull, but I notice whenever you create a new widget, there is a spinner in the header to the right of the title that comes on and eventually goes away, but I have no idea what it's spinning for. Nothing is being waiting on. The result count is good, the hints are good, etc. Maybe that spinner should go away?

rfrenchseti reviewed (changes requested):
One more bug:
- The dropdown boxes are using the "internal" name for each unit (like "saturnradii") not the "display" name (like "Rs (71492)"). See opus_support.py for how the display names are encoded. The idea is the display name should show up to the user, but when the user selects it the internal name should be put in the unit-XXX slug.

juzen2003 committed:
Update class/function/variable names and clean up code styles based on
github comments.
- In slug.py (line 73-74), add recyclebin and url_cols in SLUGS_NOT_IN_DB.
- In widget.html (line 16), change unit select to have class op-unit-*.
- In hash.js (line 444-465), change syncUpWithOPUSSelectionsForWidgetsWithNoQtype
  to syncUpOPUSSelectionsForWidgetsWithNoQtypeAndUnit and update comments
  for the function.
- In hash.js (line 174-176), update the logic in getHashStrFromSelections.
- In search.js  (line 370-400), clean up indentation for switch block.
- Change class widget-main to op-widget-main.
- Change class unit-slug to op-unit-slug.
- Change function syncUpWithOPUSSelectionsForWidgetsWithNoQtype to
  syncUpOPUSSelectionsForWidgetsWithNoQtypeAndUnit.

juzen2003 committed:
Update display names in unit select tag.
- In opus_support.py (line 540-550), add get_display_names to get a dictionary
  with valid units as keys and display names as values.
- In widget.html (line 17-18), update unit select tag to use display names
  as the text.
- In view.py (line 321), set valid_units from get_display_names.

juzen2003 committed:
Update preprogrammed ranges info when unit is changed.
- In opus_support.py (line 540), add comments for get_display_names.
- In widget.html (line 73-89), add all preprogrammed values for different
  units in html.
- In view.py (line 321-342), create a list that every item is a tuple
  of unit, minVal, maxVal. Store the list in ranges passed to widget.html.
- In opus.css (line 2287-2291), add op-hide-different-units-info to hide
  preprogrammed ranges info not for current selected unit.
- In search.js (line 422-430), add logic in select change event handler
  to update values for different units in preprogrammed ranges.
- In widgets.js (line 995-1003), add logic in getWidget done callback
  function to hide values for different units (not currently selected unit)
  in preprogrammed ranges.

juzen2003 committed:
Update param_info_ranges.json to remove unit in the header of preprogrammed
ranges.

rfrenchseti reviewed (changes requested):
- When you change the units, the URL is updated with the current input values even if they are illegal. For example, type something bad in the ring radius "min" field. Exit the field and change the widget's unit to something else. Notice that the bad value is now in the URL.
- Look at the ring radius "thebe extension" when using "m" units and narrow the browser. The numbers are wide enough that they lay on top of the word "extension".
- How are you figuring out the number of decimal places to put when you change the units in the range dropdown? For example, in the wavelength widget, select "cm" and then type in a color name. It will just show "0.0" because there aren't enough digits available. Check out the function `adjust_format_string_for_units` in opus_support.py. It will automatically adjust a format string to the proper number of decimal places given the new units.

juzen2003 committed:
Fix the issue that preprogrammed ranges collapsible will not open if
current input value has a match but previous input value has no match.
- In search.js (line 608-612, 623-627), update inputsRangesNameMatchedInfo
  to make sure all behaviors in addPreprogrammedRangesSearchBehaviors are
  correct when there is a collapsible show/hide event in compareInputWithRangesInfo.
- In search.js (line 232-233, 585-586), update selector for
  .op-preprogrammed-ranges-data-item to only select display values for
  current unit in preprogrammed ranges.

juzen2003 committed:
Update the styling for preprogrammed ranges, and use adjust_format_string_for_units
to determine the number of decimal places after a unit change.
- In opus_support.py (line 466-467), fix the display name of saturnradii
  and jupiterradii.
- In widget.html (line 78-83), add text-right to set text-align: right
  for values of preprogrammed ranges info.
- In views.py (line 334-345), use adjust_format_string_for_units to determine
  the number of decimal places after a unit change.
- In opus.css (line 2303-2306), add styling to make sure each digit has
  the same size in preprogrammed ranges.
- In search.js (line 431), remove alignRangesDataByDecimalPoint.
- In widgets.js (line 1005), remove alignRangesDataByDecimalPoint.
- In param_info_ranges.json, add "format" as the key to store the number
  of decimal places for default unit for each category.

juzen2003 committed:
Update getHashStrFromSelections function to make sure only selections with
valid values (for RANGE) get pushed to the hash string.
- In hash.js (line 79-106), modify getHashStrFromSelections to update
  ranges selections to the hash string if selections are valid.
- In widgets.js (line 477), remove debugging code.

juzen2003 committed:
Tune the styling for preprogrammed ranges when there are long numbers.
- In opus.css (line 2097-2131), update styling for preprogrammed ranges.
- In hash.js (line 90), remove debugging console.log.

juzen2003 committed:
Clean up unwanted codes and comments.
- In opus.css/search.js, clean up commented out codes.
- In widgets.js, remove alignRangesDataByDecimalPoint, getPaddingValFromDigitsInFraction.

juzen2003 committed:
Clean up commented out codes in api_get_widget in views.py.

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -603,6 +605,11 @@ var o_search = {
                     o_search.rangesNameMatchedCounterByCategory[collapsibleContainerId] += 1;
                     if (!$(`#${collapsibleContainerId}`).hasClass("show")) {
+                        // Normally inputsRangesNameMatchedInfo gets updated later in this function,
+                        // but since we are opening the collpasible item here, to make sure all behaviors
```
collapsible spelled wrong here and below

rfrenchseti commented on lib/opus_support.py:
```diff
@@ -463,8 +463,8 @@ def _range_time_decode(iso):
                 'm':            ('m', 1e-3),
-                'saturnradii':  ('Rj (60330)', 60330.),
-                'jupiterradii': ('Rs (71492)', 71492.),
+                'saturnradii':  ('Rs (60330)', 60330.),
```
Whoops. Good catch. But Jupiter should come before Saturn in the list. They are in order inner to outer planet.

rfrenchseti reviewed (changes requested):
- normalizeinput does not have the units passed, so the returned "pretty" value can have the wrong number of digits, resulting in the number being truncated incorrectly.

juzen2003 committed:
Modify url_to_search_params to have format updated from adjust_format_string_for_units
when pretty_results is true. This will make sure normalize input with unit
passed in can get the correct number of decimal places displayed.
- In opus_support.py (line 466,467), swap the order of jupiterradii and
  saturnradii in the list.
- In views.py (line 628-634), modify url_to_search_params to have format
  updated from adjust_format_string_for_units when pretty_results is true.
- In hash.js (line 93-94, 143-144), modify getHashStrFromSelections to
  set qtype and unit slug with unique input id when useFieldUniqueIDs is
  true. This is to construct the hash for normalize input api call.
- In search.js (line 162-167, 277-282), pass in unit when calling normalize
  input api. This will make sure the return values have the correct format.

juzen2003 committed:
Remove updateHashFromExtras in hash.js since the function is not used
anymore.

juzen2003 committed:
Clean up codes in views.py and add test cases for normalize input api and
url_to_search_params.
- In views.py, remove debugging print.
- In test_search.py (line 848-881), add test cases for url_to_search_params
  with unit passed in.
- In test_search_api.py (line 359-376), add test cases for normalize input
  api with unit passed in.

juzen2003 committed:
Remove uniqueid2 (at line 82) in hash.js, we will use uniqueid1 because
uniqueid1 & uniqueid2 are the same for the same input set. Fix comments
at (line 162,277) in search.js.

juzen2003 committed:
Update url_to_search_params to have sourceunit as the slug used to convert
values based on unit-slug when sourceunit is passed in. This will make sure
when normalize input api url has sourceunit-slug, all the return values
will be converted based on current selected unit in unit-slug.
- In views.py (line 489-497,524,534,581-602,690-696), add sourceunit as
  the slug used to convert values based on unit in unit-slug.
- In hash.js (line 81-87), rename uniqueid1 to uniqueid.
- In opus.js (line 63-64), create unitFromLastSearchBySlug to record the
  unit selections for widgets from last search.
- In search.js (line 445-516), update unit select change event handler
  function. When input is not empty and there is a unit change, run
  normalize input api with sourceunit-slug (value: previous selected unit)
  on all inputs in the same widget. The api return values will be properly
  converted based on current selected unit, and we will update all inputs
  with converted return values.

juzen2003 committed:
Add test cases for normalize input api with unit & sourceunit. Add comments
in search.js.
- In search.js (line 447-449), add comments.
- In test_search_api.py (line 377-399), add tests for normalize input api
  with unit and sourceunit.

juzen2003 committed:
Init opus.unitFromLastSearchBySlug when a widget with unit select is open.
- In widgets.js (line 936), Init opus.unitFromLastSearchBySlug when a
  widget with unit select is open.

rfrenchseti commented on opus/application/apps/search/views.py:
```diff
@@ -653,6 +687,13 @@ def url_to_search_params(request_get, allow_errors=False, return_slugs=False,
                                       +'function "%s"', form_type_func)
                             return None, None
                     if value_to_use:
+                        if is_sourceunit or sourceunit_val is not None:
```
Note the code starting at line 697. It catches exceptions during the execution of func() and prevents the back end from crashing. The new code calls func() when there’s a sourceunit without catching these. Also if there’s a sourceunit, func() gets called twice, once here and once inside the try. Not fatal if func is float or int, but bad if it’s something converting with a more complicated format. 

As much as I hate putting lots of code in a try block, I think the right thing to do is move the unit conversion a few lines down to just before the if pretty_results and use new_value instead of func(value_to_use).

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -376,6 +376,8 @@ var o_search = {
         // range behaviors and string behaviors for search widgets - qtype select dropdown
         $('#search').on("change", "select", function() {
             let isInputSetEmpty = true;
+            // Use this flag to determine if
```
Comment is truncated

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -376,6 +376,8 @@ var o_search = {
             let isInputSetEmpty = true;
+            // Use this flag to determine if
+            let performNormalizeInput = false;
             if ($(this).attr("name").startsWith("qtype-")) {
```
The comment before this function should mention unit drop down too

rfrenchseti reviewed (changes requested):
- There's a race condition with the new unit conversion. If you change units from A to B, and then change units again B to C before the first normalizeinput returns, the second call to normalizeinput gets called with unit=C, sourceunit=B but the numbers that get passed in are still in units of A. I'm not sure what the right way to fix this is. It seems like you should keep the units along with the numbers in the inputs, not just in the dropdown box. Then update the units in the inputs when normalizeinput successfully returns and you're updating the input contents.

juzen2003 committed:
Rename unitFromLastSearchBySlug to currentUnitBySlug, and update
currentUnitBySlug at the return of normalize input api call so that the
unit in currentUnitBySlug will match the values from normalize input api
return. This will solve the race condition of unit conversions when
user quickly switches to another different unit before the return of
normalize input api call from previous unit selection.
- In views.py (line 696-703), move this block of codes inside the try block.
- In search.js (line 860-864), update currentUnitBySlug at the return of
  normalize input api call.
- In hash.js/widgets.js/search.js/opus.js, rename unitFromLastSearchBySlug
  to currentUnitBySlug

rfrenchseti reviewed (changes requested):
- Type in a value. Change the unit without otherwise leaving focus first. The value you just typed will not be converted.
- Something similar happens with preprogrammed ranges. Type in a range name. Click on the range. Change the units before normalizeinput returns. The value will be in the wrong units.

juzen2003 committed:
Update unit select change event handler to use the actual input elements'
values to create the hash for normalize input api. This will make sure the
latest input values are used for normalize input api url. Previously we
used  values in opus.selections, but those values could be outdated if
opus.selections is still waiting for the return of an early normalize
input api.
- In search.js (line 473-486), use the actual input elements' values to
  create the hash for normalize input api in unit select change event
  handler.
- In search.js (line 533-572), create areAllInputsValInAWidgetEmpty to check
  if the actual input elements' values are empty in a widget.

juzen2003 committed:
Fix the issue that range input background doesn't turn from red to white
when its value has changed from an invalid value to a valid value.
- In search.js (line 880-883), properly clear the border of an input if
  normalize input api return value for that input is valid.

juzen2003 committed:
Fix the page reload issue when string input has "," in search value.
- In hash.js (line 453), in getSelectionsExtrasFromHash, when there is no
  slugCounter, we know there is only one selection for the slug and the
  selection might contains ",". Instead of using value.split(","), we use
  [value] to create an array of one element and make sure "," is included.
- In hash.js (line 655), in initFromHash, when slug is not in opus.prefs
  and there is no slugCounter, we know there is only one selection. we use
  [value] to create an array of one element in opus.selections and make
  sure "," is included.
- In hash.js (line 304-309, 354-359), when slug is not in opus.prefs, we
  use [value] to create an array of one element in and make sure "," is
  included.
- In opus.js (line 216), remove an extra empty line.

juzen2003 committed:
Make selections for mults to be an array of one element, and the value
will be a string that joins multiple search values by comma.
- In search.js (line 350, 360-367), in choices input change event handler,
  make the value as a string that joins multiple search values by comma,
  and update opus.selections to be an arry of one element ([value]).
- In search.js (line 1169), modify clearInputBorder to take in input tag
  as the parameter, instead of input set as the parameter.
- In search.js (line 884), replace the code block with clearInputBorder.
- In widgets.js (line 166), pass cloneInputs.find("input") to clearInputBorder
  since clearInputBorder now takes in input tag.

juzen2003 committed:
Update encodeSlugValues to not encode "," to "%2C".
In hash.js (line 281), modify encodeSlugValues to not encode "," to "%2C".

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 46_units_selection
